### PR TITLE
blog: Debugging qemu with gdb

### PR DESCRIPTION
**Pull request by kallisti5**

### CODE REVIEW

File: content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+### Compiling Haiku (with debug symbols)
+
+Overall, this is the easier step for most folks with any Haiku development experience.
+
+1. Get the [haiku and buildtools git repos](/guides/building/get-source-git).
+2. Compile your toolchain.
+   1. ```cd haiku; mkdir generated.riscv64; cd generated.riscv64```
+   1. ```../configure -j2 --cross-tools-source ../../buildtools --build-cross-tools riscv64```
+3. Enable debugging symbols in bootloader.
+   1. Copy UserBuildConfig.example to UserBuildConfig in haiku/build/jam/
```
Comment (pulkomandy) [inline content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md:33]: That's not really needed, you can start from an empty file and just add a single SetConfigVar line to it.

File: content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+### Attach GDB
+
+Now, we attach to qemu from GDB, and start debugging!
+
+*TIP: The hex addresses below could change over time. If you don't get debugging symbols (aka source-code view of what's happening) you'll need to "figure the new addresses out". I've included some basic information on how I got them.*
+
+1. Start GDB for your architecture
+   1. ```riscv64-elf-gdb /tmp/riscv64-build/u-boot``` on ArchLinux
+2. Connect to qemu
+   1. ```target remote 127.0.0.1:1234```
```
Comment (pulkomandy) [inline content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md:88]: gdb has a "target extended-remote" with some additional features, in particular it should be able to handle qemu reboot (using the gdb "run" command, I think). It makes things a bit less "one shot".

File: content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+3. Tell qemu where u-boot symbols should be loaded
```
Comment (pulkomandy) [inline content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md:89]: Actually it's "tell gdb", qemu does not care about symbols

Comment (kallisti5) [reply content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md:89]: oops. nice catch!

File: content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md (hunk 1 above)
Comment (kallisti5) [reply content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md:33]: fair.  I just normally use it as an "example"

File: content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md (hunk 2 above)
Comment (kallisti5) [reply content/blog/kallisti5/2021-01-17_debugging_riscv-64_bootloader_in_qemu.md:88]: i tried it, ```run``` just seems to make qemu puke and exit.   I've modified remote to extended-remote and included a little blurb about the potential to debug without redoing everything every time.

**Pull request merged**